{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c5aac698-f7ed-4489-8c6c-f75ec3f3df73",
   "metadata": {},
   "source": [
    "# Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "445d6e41-c30f-4cbc-b5b3-97dc1020f09e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyaging as pya"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "895dcb1b-20db-4d91-aebf-bbdc4af14ab7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting find_clock_by_doi function\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0009s]\n",
      "|-----> ⚙️ Searching for clock based on DOI started\n",
      "|-----------> in progress: 100.0000%\n",
      "|-----------> Clocks with DOI https://doi.org/10.1038/s43587-022-00248-2: pchorvath2013, pcphenoage, pcgrimage, pchannum, pcdnamtl, pcskinandblood\n",
      "|-----> ✅ Searching for clock based on DOI finished [0.0047s]\n",
      "|-----> 🎉 Done! [0.0209s]\n"
     ]
    }
   ],
   "source": [
    "pya.utils.find_clock_by_doi('https://doi.org/10.1038/s43587-022-00248-2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3a14445f-3f19-415c-a768-4cd9d7c193fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting cite_clock function\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0010s]\n",
      "|-----> ⚙️ Searching for citation of clock pchorvath2013 started\n",
      "|-----------> Citation for pchorvath2013:\n",
      "|-----------> Higgins-Chen, Albert T., et al. \"A computational solution for bolstering reliability of epigenetic clocks: Implications for clinical trials and longitudinal tracking.\" Nature aging 2.7 (2022): 644-661.\n",
      "|-----> ✅ Searching for citation of clock pchorvath2013 finished [0.0018s]\n",
      "|-----> 🎉 Done! [0.0040s]\n"
     ]
    }
   ],
   "source": [
    "pya.utils.cite_clock('PCHorvath2013')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "13b14cee-d51b-49f1-b7c0-4d03dd53a278",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting show_all_clocks function\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0017s]\n",
      "|-----> ⚙️ Showing all available clock names started\n",
      "|-----------> horvath2013\n",
      "|-----------> h3k36me3\n",
      "|-----------> panhistone\n",
      "|-----------> ocampoatac2\n",
      "|-----------> h3k4me1\n",
      "|-----------> h3k27ac\n",
      "|-----------> mammalian1\n",
      "|-----------> pchorvath2013\n",
      "|-----------> replitali\n",
      "|-----------> h3k9me3\n",
      "|-----------> pcphenoage\n",
      "|-----------> skinandblood\n",
      "|-----------> pcgrimage\n",
      "|-----------> pchannum\n",
      "|-----------> mammalianlifespan\n",
      "|-----------> pcdnamtl\n",
      "|-----------> pedbe\n",
      "|-----------> dunedinpace\n",
      "|-----------> bitage\n",
      "|-----------> dnamtl\n",
      "|-----------> ocampoatac1\n",
      "|-----------> h3k9ac\n",
      "|-----------> h3k27me3\n",
      "|-----------> mammalian2\n",
      "|-----------> altumage\n",
      "|-----------> h3k4me3\n",
      "|-----------> phenoage\n",
      "|-----------> pcskinandblood\n",
      "|-----------> mammalian3\n",
      "|-----------> hannum\n",
      "|-----> ✅ Showing all available clock names finished [0.0024s]\n",
      "|-----> 🎉 Done! [0.0120s]\n"
     ]
    }
   ],
   "source": [
    "pya.utils.show_all_clocks()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f822e4bb-88e7-4809-808f-701185c1e5ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting get_clock_metadata function\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0015s]\n",
      "|-----> ⚙️ Showing pchorvath2013 metadata started\n",
      "|-----------> species: Homo sapiens\n",
      "|-----------> data_type: methylation\n",
      "|-----------> year: 2022\n",
      "|-----------> citation: Higgins-Chen, Albert T., et al. \"A computational solution for bolstering reliability of epigenetic clocks: Implications for clinical trials and longitudinal tracking.\" Nature aging 2.7 (2022): 644-661.\n",
      "|-----------> doi: https://doi.org/10.1038/s43587-022-00248-2\n",
      "|-----> ✅ Showing pchorvath2013 metadata finished [0.0037s]\n",
      "|-----> 🎉 Done! [0.0101s]\n"
     ]
    }
   ],
   "source": [
    "pya.utils.get_clock_metadata('PCHorvath2013')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62e1ac68-927d-4ca8-a2ab-bd99a7ee52ab",
   "metadata": {},
   "source": [
    "# Methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dc347ae0-41ae-46ac-ba50-08cacd4c9241",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyaging as pya"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d771588a-01e1-4220-90f9-3048152a7efd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting download_example_data function\n",
      "|-----------> ⚙️ Download example data started\n",
      "|-----------------> Data found in ./pyaging_data/GSE139307.pkl\n",
      "|-----------> ✅ Download example data finished [0.0027s]\n",
      "|-----> 🎉 Done! [0.0045s]\n"
     ]
    }
   ],
   "source": [
    "pya.data.download_example_data('methylation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3192d1bd-f5d6-426c-8c65-29971e46c4b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle('pyaging_data/GSE139307.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b12f7bbf-ec33-4b7a-a9f5-f06309885966",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['female'] = (df['gender'] == 'F').astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "285dec20-7b30-491a-9bda-8bc784faae6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>cg00000029</th>\n",
       "      <th>cg00000108</th>\n",
       "      <th>cg00000109</th>\n",
       "      <th>cg00000165</th>\n",
       "      <th>cg00000236</th>\n",
       "      <th>cg00000289</th>\n",
       "      <th>...</th>\n",
       "      <th>ch.X.938089F</th>\n",
       "      <th>ch.X.94051109R</th>\n",
       "      <th>ch.X.94260649R</th>\n",
       "      <th>ch.X.967194F</th>\n",
       "      <th>ch.X.97129969R</th>\n",
       "      <th>ch.X.97133160R</th>\n",
       "      <th>ch.X.97651759F</th>\n",
       "      <th>ch.X.97737721F</th>\n",
       "      <th>ch.X.98007042R</th>\n",
       "      <th>female</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM4137709</th>\n",
       "      <td>GSE139307</td>\n",
       "      <td>sperm</td>\n",
       "      <td>84.0</td>\n",
       "      <td>M</td>\n",
       "      <td>0.084811</td>\n",
       "      <td>0.920696</td>\n",
       "      <td>0.856851</td>\n",
       "      <td>0.084567</td>\n",
       "      <td>0.838699</td>\n",
       "      <td>0.247273</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045942</td>\n",
       "      <td>0.037631</td>\n",
       "      <td>0.056455</td>\n",
       "      <td>0.249872</td>\n",
       "      <td>0.049022</td>\n",
       "      <td>0.085691</td>\n",
       "      <td>0.037435</td>\n",
       "      <td>0.077820</td>\n",
       "      <td>0.106234</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137710</th>\n",
       "      <td>GSE139307</td>\n",
       "      <td>sperm</td>\n",
       "      <td>69.0</td>\n",
       "      <td>M</td>\n",
       "      <td>0.099626</td>\n",
       "      <td>0.919073</td>\n",
       "      <td>0.890024</td>\n",
       "      <td>0.115541</td>\n",
       "      <td>0.852584</td>\n",
       "      <td>0.198103</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041849</td>\n",
       "      <td>0.032573</td>\n",
       "      <td>0.089790</td>\n",
       "      <td>0.250245</td>\n",
       "      <td>0.079095</td>\n",
       "      <td>0.079756</td>\n",
       "      <td>0.046229</td>\n",
       "      <td>0.091256</td>\n",
       "      <td>0.120241</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137711</th>\n",
       "      <td>GSE139307</td>\n",
       "      <td>sperm</td>\n",
       "      <td>69.0</td>\n",
       "      <td>M</td>\n",
       "      <td>0.117228</td>\n",
       "      <td>0.920276</td>\n",
       "      <td>0.894317</td>\n",
       "      <td>0.117127</td>\n",
       "      <td>0.839258</td>\n",
       "      <td>0.213410</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049515</td>\n",
       "      <td>0.058097</td>\n",
       "      <td>0.079919</td>\n",
       "      <td>0.299758</td>\n",
       "      <td>0.079305</td>\n",
       "      <td>0.089815</td>\n",
       "      <td>0.065364</td>\n",
       "      <td>0.086864</td>\n",
       "      <td>0.156005</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137712</th>\n",
       "      <td>GSE139307</td>\n",
       "      <td>sperm</td>\n",
       "      <td>69.0</td>\n",
       "      <td>M</td>\n",
       "      <td>0.077096</td>\n",
       "      <td>0.910204</td>\n",
       "      <td>0.908400</td>\n",
       "      <td>0.073885</td>\n",
       "      <td>0.861615</td>\n",
       "      <td>0.163276</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033289</td>\n",
       "      <td>0.038836</td>\n",
       "      <td>0.108213</td>\n",
       "      <td>0.295428</td>\n",
       "      <td>0.050731</td>\n",
       "      <td>0.099943</td>\n",
       "      <td>0.047597</td>\n",
       "      <td>0.078480</td>\n",
       "      <td>0.107480</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137713</th>\n",
       "      <td>GSE139307</td>\n",
       "      <td>sperm</td>\n",
       "      <td>67.0</td>\n",
       "      <td>M</td>\n",
       "      <td>0.063524</td>\n",
       "      <td>0.911608</td>\n",
       "      <td>0.884643</td>\n",
       "      <td>0.079877</td>\n",
       "      <td>0.864654</td>\n",
       "      <td>0.176169</td>\n",
       "      <td>...</td>\n",
       "      <td>0.038411</td>\n",
       "      <td>0.048787</td>\n",
       "      <td>0.088631</td>\n",
       "      <td>0.316694</td>\n",
       "      <td>0.041873</td>\n",
       "      <td>0.079303</td>\n",
       "      <td>0.048823</td>\n",
       "      <td>0.089010</td>\n",
       "      <td>0.117903</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 485517 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              dataset tissue_type   age gender  cg00000029  cg00000108  \\\n",
       "GSM4137709  GSE139307       sperm  84.0      M    0.084811    0.920696   \n",
       "GSM4137710  GSE139307       sperm  69.0      M    0.099626    0.919073   \n",
       "GSM4137711  GSE139307       sperm  69.0      M    0.117228    0.920276   \n",
       "GSM4137712  GSE139307       sperm  69.0      M    0.077096    0.910204   \n",
       "GSM4137713  GSE139307       sperm  67.0      M    0.063524    0.911608   \n",
       "\n",
       "            cg00000109  cg00000165  cg00000236  cg00000289  ...  ch.X.938089F  \\\n",
       "GSM4137709    0.856851    0.084567    0.838699    0.247273  ...      0.045942   \n",
       "GSM4137710    0.890024    0.115541    0.852584    0.198103  ...      0.041849   \n",
       "GSM4137711    0.894317    0.117127    0.839258    0.213410  ...      0.049515   \n",
       "GSM4137712    0.908400    0.073885    0.861615    0.163276  ...      0.033289   \n",
       "GSM4137713    0.884643    0.079877    0.864654    0.176169  ...      0.038411   \n",
       "\n",
       "            ch.X.94051109R  ch.X.94260649R  ch.X.967194F  ch.X.97129969R  \\\n",
       "GSM4137709        0.037631        0.056455      0.249872        0.049022   \n",
       "GSM4137710        0.032573        0.089790      0.250245        0.079095   \n",
       "GSM4137711        0.058097        0.079919      0.299758        0.079305   \n",
       "GSM4137712        0.038836        0.108213      0.295428        0.050731   \n",
       "GSM4137713        0.048787        0.088631      0.316694        0.041873   \n",
       "\n",
       "            ch.X.97133160R  ch.X.97651759F  ch.X.97737721F  ch.X.98007042R  \\\n",
       "GSM4137709        0.085691        0.037435        0.077820        0.106234   \n",
       "GSM4137710        0.079756        0.046229        0.091256        0.120241   \n",
       "GSM4137711        0.089815        0.065364        0.086864        0.156005   \n",
       "GSM4137712        0.099943        0.047597        0.078480        0.107480   \n",
       "GSM4137713        0.079303        0.048823        0.089010        0.117903   \n",
       "\n",
       "            female  \n",
       "GSM4137709       0  \n",
       "GSM4137710       0  \n",
       "GSM4137711       0  \n",
       "GSM4137712       0  \n",
       "GSM4137713       0  \n",
       "\n",
       "[5 rows x 485517 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "671906c4-ae7e-4808-859a-470e19757477",
   "metadata": {},
   "outputs": [],
   "source": [
    "# needs only numerical data (doesn't work with strings)\n",
    "df = df.drop(['gender', 'tissue_type', 'dataset'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6f68ea1-dbbb-4a21-8b2b-f36a7df47222",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting df_to_adata function\n",
      "|-----> ⚙️ Impute missing values started\n",
      "|-----------> Imputing missing values using mean strategy\n",
      "|-----> ✅ Impute missing values finished [0.1147s]\n",
      "|-----> ⚙️ Log data statistics started\n",
      "|-----------> There are 37 observations\n",
      "|-----------> There are 485514 features\n",
      "|-----------> Total missing values: 0\n",
      "|-----------> Percentage of missing values: 0.00%\n",
      "|-----> ✅ Log data statistics finished [0.2208s]\n",
      "|-----> ⚙️ Create anndata object started\n",
      "|-----> ✅ Create anndata object finished [0.0169s]\n",
      "|-----> ⚙️ Add metadata to anndata started\n",
      "|-----------? No metadata provided. Leaving adata.obs empty\n",
      "|-----> ⚠️ Add metadata to anndata finished [0.0336s]\n",
      "|-----> ⚙️ Add unstructured data to anndata started\n",
      "|-----> ✅ Add unstructured data to anndata finished [0.0009s]\n",
      "|-----> 🎉 Done! [0.3873s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.pp.df_to_adata(df, imputer_strategy='mean')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e4de7a8-0898-4cbe-89e6-69d48ccd039a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting predict_age function\n",
      "|-----> ⚙️ Set PyTorch device started\n",
      "|-----------> Using device: cpu\n",
      "|-----> ✅ Set PyTorch device finished [0.0009s]\n",
      "|-----> Processing clock: altumage\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/altumage.pt\n",
      "|-----------> ✅ Load clock finished [0.0033s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.0107s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0167s]\n",
      "|-----------> ⚙️ Preprocess data started\n",
      "|-----------------> Preprocessing data with function scale\n",
      "|-----------> ✅ Preprocess data finished [0.0162s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0048s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0064s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0057s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0011s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0022s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0034s]\n",
      "|-----> Processing clock: horvath2013\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/horvath2013.pt\n",
      "|-----------> ✅ Load clock finished [0.0020s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.0023s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0017s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0022s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0012s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0014s]\n",
      "|-----------> ⚙️ Postprocess data started\n",
      "|-----------------> Postprocessing data with function anti_log_linear\n",
      "|-----------> ✅ Postprocess data finished [0.0007s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0019s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0014s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0024s]\n",
      "|-----> Processing clock: pchorvath2013\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/pchorvath2013.pt\n",
      "|-----------> ✅ Load clock finished [0.0011s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.0330s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0478s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0357s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0107s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0094s]\n",
      "|-----------> ⚙️ Postprocess data started\n",
      "|-----------------> Postprocessing data with function anti_log_linear\n",
      "|-----------> ✅ Postprocess data finished [0.0011s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0014s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0012s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0021s]\n",
      "|-----> Processing clock: dunedinpace\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/dunedinpace.pt\n",
      "|-----------> ✅ Load clock finished [0.0010s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.0052s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0013s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0018s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0019s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0010s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0007s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0015s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0018s]\n",
      "|-----> Processing clock: replitali\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/replitali.pt\n",
      "|-----------> ✅ Load clock finished [0.0020s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------------? 57 out of 87 features (65.52%) are missing and will be added with default value 0: ['cg00077044', 'cg00150168', 'cg00454443'], etc.\n",
      "|-----------------> Expanded adata with 57 missing features.\n",
      "|-----------> ⚠️ Check features in adata finished [0.0022s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.1583s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0054s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0014s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0013s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0008s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0010s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0018s]\n",
      "|-----> Processing clock: skinandblood\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/skinandblood.pt\n",
      "|-----------> ✅ Load clock finished [0.0014s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.0035s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0027s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0021s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0016s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0013s]\n",
      "|-----------> ⚙️ Postprocess data started\n",
      "|-----------------> Postprocessing data with function anti_log_linear\n",
      "|-----------> ✅ Postprocess data finished [0.0009s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0012s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0009s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0054s]\n",
      "|-----> Processing clock: mammalian2\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Data found in ./pyaging_data/mammalian2.pt\n",
      "|-----------> ✅ Load clock finished [0.0026s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------------? 680 out of 816 features (83.33%) are missing and will be added with default value 0: ['cg00020468', 'cg00098422', 'cg00106940'], etc.\n",
      "|-----------------> Expanded adata with 680 missing features.\n",
      "|-----------> ⚠️ Check features in adata finished [0.0090s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.1685s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0079s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0011s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0009s]\n",
      "|-----------> ⚙️ Postprocess data started\n",
      "|-----------------> Postprocessing data with function anti_log_log\n",
      "|-----------> ✅ Postprocess data finished [0.0008s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0009s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0010s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0019s]\n",
      "|-----> 🎉 Done! [0.6389s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.pred.predict_age(adata, ['altumage', 'horvath2013', 'pchorvath2013', 'dunedinpace', 'replitali', 'skinandblood', 'mammalian2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9c0c6926-2944-4274-aefa-eb099b8e6737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>altumage</th>\n",
       "      <th>horvath2013</th>\n",
       "      <th>pchorvath2013</th>\n",
       "      <th>dunedinpace</th>\n",
       "      <th>replitali</th>\n",
       "      <th>skinandblood</th>\n",
       "      <th>mammalian2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM4137709</th>\n",
       "      <td>37.007023</td>\n",
       "      <td>33.624774</td>\n",
       "      <td>4.672973</td>\n",
       "      <td>1.152685</td>\n",
       "      <td>87.261047</td>\n",
       "      <td>2.354882</td>\n",
       "      <td>0.007100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137710</th>\n",
       "      <td>29.426702</td>\n",
       "      <td>28.829343</td>\n",
       "      <td>2.086035</td>\n",
       "      <td>1.031971</td>\n",
       "      <td>87.671936</td>\n",
       "      <td>1.060467</td>\n",
       "      <td>0.007577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137711</th>\n",
       "      <td>22.805416</td>\n",
       "      <td>28.316548</td>\n",
       "      <td>2.549487</td>\n",
       "      <td>1.063786</td>\n",
       "      <td>86.664955</td>\n",
       "      <td>1.289146</td>\n",
       "      <td>0.003947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137712</th>\n",
       "      <td>18.060070</td>\n",
       "      <td>24.850635</td>\n",
       "      <td>1.164793</td>\n",
       "      <td>1.142039</td>\n",
       "      <td>86.230934</td>\n",
       "      <td>0.595460</td>\n",
       "      <td>0.003872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137713</th>\n",
       "      <td>20.071888</td>\n",
       "      <td>25.942114</td>\n",
       "      <td>1.491444</td>\n",
       "      <td>1.115847</td>\n",
       "      <td>87.336739</td>\n",
       "      <td>0.786787</td>\n",
       "      <td>0.004309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137714</th>\n",
       "      <td>15.904675</td>\n",
       "      <td>27.717805</td>\n",
       "      <td>1.116081</td>\n",
       "      <td>1.180822</td>\n",
       "      <td>86.192078</td>\n",
       "      <td>0.509315</td>\n",
       "      <td>0.003994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137715</th>\n",
       "      <td>17.618645</td>\n",
       "      <td>25.023288</td>\n",
       "      <td>1.132691</td>\n",
       "      <td>1.120775</td>\n",
       "      <td>87.315369</td>\n",
       "      <td>0.786464</td>\n",
       "      <td>0.004123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137716</th>\n",
       "      <td>18.396324</td>\n",
       "      <td>26.629503</td>\n",
       "      <td>1.122279</td>\n",
       "      <td>1.201142</td>\n",
       "      <td>86.422653</td>\n",
       "      <td>0.336474</td>\n",
       "      <td>0.004455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137717</th>\n",
       "      <td>25.673313</td>\n",
       "      <td>27.471301</td>\n",
       "      <td>1.180770</td>\n",
       "      <td>1.208640</td>\n",
       "      <td>86.630707</td>\n",
       "      <td>0.562296</td>\n",
       "      <td>0.004236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137718</th>\n",
       "      <td>15.824392</td>\n",
       "      <td>26.320859</td>\n",
       "      <td>1.132404</td>\n",
       "      <td>1.126852</td>\n",
       "      <td>86.628609</td>\n",
       "      <td>0.730006</td>\n",
       "      <td>0.004447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137719</th>\n",
       "      <td>38.558460</td>\n",
       "      <td>32.887796</td>\n",
       "      <td>5.543674</td>\n",
       "      <td>1.034858</td>\n",
       "      <td>87.878998</td>\n",
       "      <td>2.630242</td>\n",
       "      <td>0.005451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137720</th>\n",
       "      <td>20.418158</td>\n",
       "      <td>27.964102</td>\n",
       "      <td>1.353481</td>\n",
       "      <td>1.209375</td>\n",
       "      <td>87.149872</td>\n",
       "      <td>0.525785</td>\n",
       "      <td>0.004212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137721</th>\n",
       "      <td>19.979923</td>\n",
       "      <td>26.013752</td>\n",
       "      <td>1.432106</td>\n",
       "      <td>1.189186</td>\n",
       "      <td>86.588623</td>\n",
       "      <td>0.459680</td>\n",
       "      <td>0.002872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137722</th>\n",
       "      <td>24.194540</td>\n",
       "      <td>28.065958</td>\n",
       "      <td>1.300553</td>\n",
       "      <td>1.093692</td>\n",
       "      <td>87.032524</td>\n",
       "      <td>0.604637</td>\n",
       "      <td>0.004160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137723</th>\n",
       "      <td>17.190454</td>\n",
       "      <td>27.533139</td>\n",
       "      <td>1.779086</td>\n",
       "      <td>1.188542</td>\n",
       "      <td>86.472771</td>\n",
       "      <td>0.853337</td>\n",
       "      <td>0.003363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137724</th>\n",
       "      <td>18.759686</td>\n",
       "      <td>24.812141</td>\n",
       "      <td>1.099413</td>\n",
       "      <td>1.194590</td>\n",
       "      <td>85.776878</td>\n",
       "      <td>0.352591</td>\n",
       "      <td>0.004083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137725</th>\n",
       "      <td>17.183453</td>\n",
       "      <td>24.013267</td>\n",
       "      <td>0.936747</td>\n",
       "      <td>1.124992</td>\n",
       "      <td>86.564934</td>\n",
       "      <td>0.660432</td>\n",
       "      <td>0.003064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137726</th>\n",
       "      <td>18.004549</td>\n",
       "      <td>26.001558</td>\n",
       "      <td>1.087593</td>\n",
       "      <td>1.177684</td>\n",
       "      <td>85.883621</td>\n",
       "      <td>0.449509</td>\n",
       "      <td>0.003430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137727</th>\n",
       "      <td>20.223724</td>\n",
       "      <td>22.912350</td>\n",
       "      <td>1.334903</td>\n",
       "      <td>1.309424</td>\n",
       "      <td>85.873466</td>\n",
       "      <td>0.565050</td>\n",
       "      <td>0.004039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137728</th>\n",
       "      <td>19.228153</td>\n",
       "      <td>29.268841</td>\n",
       "      <td>1.366070</td>\n",
       "      <td>1.157657</td>\n",
       "      <td>87.335876</td>\n",
       "      <td>0.385037</td>\n",
       "      <td>0.005384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137729</th>\n",
       "      <td>19.577639</td>\n",
       "      <td>28.927794</td>\n",
       "      <td>1.165986</td>\n",
       "      <td>1.217237</td>\n",
       "      <td>87.535812</td>\n",
       "      <td>0.580252</td>\n",
       "      <td>0.005072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137730</th>\n",
       "      <td>45.032902</td>\n",
       "      <td>35.591832</td>\n",
       "      <td>3.758405</td>\n",
       "      <td>1.110375</td>\n",
       "      <td>88.962883</td>\n",
       "      <td>1.592233</td>\n",
       "      <td>0.008719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137731</th>\n",
       "      <td>19.191566</td>\n",
       "      <td>30.693108</td>\n",
       "      <td>1.637652</td>\n",
       "      <td>1.045655</td>\n",
       "      <td>87.536636</td>\n",
       "      <td>0.346334</td>\n",
       "      <td>0.004748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137732</th>\n",
       "      <td>14.982388</td>\n",
       "      <td>25.208827</td>\n",
       "      <td>1.184326</td>\n",
       "      <td>1.207765</td>\n",
       "      <td>85.309921</td>\n",
       "      <td>0.533043</td>\n",
       "      <td>0.003144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137733</th>\n",
       "      <td>16.940281</td>\n",
       "      <td>25.472710</td>\n",
       "      <td>1.049112</td>\n",
       "      <td>1.200478</td>\n",
       "      <td>86.058090</td>\n",
       "      <td>0.343414</td>\n",
       "      <td>0.003209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137734</th>\n",
       "      <td>18.857653</td>\n",
       "      <td>25.281781</td>\n",
       "      <td>1.264270</td>\n",
       "      <td>1.153367</td>\n",
       "      <td>85.981239</td>\n",
       "      <td>0.417019</td>\n",
       "      <td>0.004968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137735</th>\n",
       "      <td>18.440538</td>\n",
       "      <td>21.608184</td>\n",
       "      <td>0.809690</td>\n",
       "      <td>0.976271</td>\n",
       "      <td>85.094543</td>\n",
       "      <td>0.386166</td>\n",
       "      <td>0.004448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137736</th>\n",
       "      <td>21.025726</td>\n",
       "      <td>26.555988</td>\n",
       "      <td>1.246243</td>\n",
       "      <td>1.177747</td>\n",
       "      <td>87.425606</td>\n",
       "      <td>0.578503</td>\n",
       "      <td>0.003434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137737</th>\n",
       "      <td>20.404150</td>\n",
       "      <td>23.738753</td>\n",
       "      <td>1.298100</td>\n",
       "      <td>1.171406</td>\n",
       "      <td>87.339378</td>\n",
       "      <td>0.717524</td>\n",
       "      <td>0.006604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137738</th>\n",
       "      <td>19.963135</td>\n",
       "      <td>21.407858</td>\n",
       "      <td>0.956814</td>\n",
       "      <td>1.199306</td>\n",
       "      <td>86.457207</td>\n",
       "      <td>0.574664</td>\n",
       "      <td>0.003424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137739</th>\n",
       "      <td>30.701977</td>\n",
       "      <td>25.379519</td>\n",
       "      <td>1.268822</td>\n",
       "      <td>1.193359</td>\n",
       "      <td>88.006493</td>\n",
       "      <td>0.663944</td>\n",
       "      <td>0.003469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137740</th>\n",
       "      <td>19.709518</td>\n",
       "      <td>26.994544</td>\n",
       "      <td>1.140624</td>\n",
       "      <td>1.174096</td>\n",
       "      <td>87.551102</td>\n",
       "      <td>0.629927</td>\n",
       "      <td>0.004361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137741</th>\n",
       "      <td>25.006363</td>\n",
       "      <td>29.815408</td>\n",
       "      <td>2.834759</td>\n",
       "      <td>1.106260</td>\n",
       "      <td>86.890930</td>\n",
       "      <td>1.303225</td>\n",
       "      <td>0.004899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137742</th>\n",
       "      <td>19.827387</td>\n",
       "      <td>29.370889</td>\n",
       "      <td>0.959647</td>\n",
       "      <td>1.257892</td>\n",
       "      <td>86.666138</td>\n",
       "      <td>0.374429</td>\n",
       "      <td>0.004270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137743</th>\n",
       "      <td>16.928745</td>\n",
       "      <td>27.138979</td>\n",
       "      <td>1.100993</td>\n",
       "      <td>1.179078</td>\n",
       "      <td>87.818886</td>\n",
       "      <td>0.499079</td>\n",
       "      <td>0.004071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137744</th>\n",
       "      <td>16.565306</td>\n",
       "      <td>23.659621</td>\n",
       "      <td>1.083474</td>\n",
       "      <td>1.186454</td>\n",
       "      <td>86.381805</td>\n",
       "      <td>0.481675</td>\n",
       "      <td>0.003689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4137745</th>\n",
       "      <td>20.054005</td>\n",
       "      <td>31.345807</td>\n",
       "      <td>1.403846</td>\n",
       "      <td>1.169119</td>\n",
       "      <td>86.564972</td>\n",
       "      <td>0.473375</td>\n",
       "      <td>0.005378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             altumage  horvath2013  pchorvath2013  dunedinpace  replitali  \\\n",
       "GSM4137709  37.007023    33.624774       4.672973     1.152685  87.261047   \n",
       "GSM4137710  29.426702    28.829343       2.086035     1.031971  87.671936   \n",
       "GSM4137711  22.805416    28.316548       2.549487     1.063786  86.664955   \n",
       "GSM4137712  18.060070    24.850635       1.164793     1.142039  86.230934   \n",
       "GSM4137713  20.071888    25.942114       1.491444     1.115847  87.336739   \n",
       "GSM4137714  15.904675    27.717805       1.116081     1.180822  86.192078   \n",
       "GSM4137715  17.618645    25.023288       1.132691     1.120775  87.315369   \n",
       "GSM4137716  18.396324    26.629503       1.122279     1.201142  86.422653   \n",
       "GSM4137717  25.673313    27.471301       1.180770     1.208640  86.630707   \n",
       "GSM4137718  15.824392    26.320859       1.132404     1.126852  86.628609   \n",
       "GSM4137719  38.558460    32.887796       5.543674     1.034858  87.878998   \n",
       "GSM4137720  20.418158    27.964102       1.353481     1.209375  87.149872   \n",
       "GSM4137721  19.979923    26.013752       1.432106     1.189186  86.588623   \n",
       "GSM4137722  24.194540    28.065958       1.300553     1.093692  87.032524   \n",
       "GSM4137723  17.190454    27.533139       1.779086     1.188542  86.472771   \n",
       "GSM4137724  18.759686    24.812141       1.099413     1.194590  85.776878   \n",
       "GSM4137725  17.183453    24.013267       0.936747     1.124992  86.564934   \n",
       "GSM4137726  18.004549    26.001558       1.087593     1.177684  85.883621   \n",
       "GSM4137727  20.223724    22.912350       1.334903     1.309424  85.873466   \n",
       "GSM4137728  19.228153    29.268841       1.366070     1.157657  87.335876   \n",
       "GSM4137729  19.577639    28.927794       1.165986     1.217237  87.535812   \n",
       "GSM4137730  45.032902    35.591832       3.758405     1.110375  88.962883   \n",
       "GSM4137731  19.191566    30.693108       1.637652     1.045655  87.536636   \n",
       "GSM4137732  14.982388    25.208827       1.184326     1.207765  85.309921   \n",
       "GSM4137733  16.940281    25.472710       1.049112     1.200478  86.058090   \n",
       "GSM4137734  18.857653    25.281781       1.264270     1.153367  85.981239   \n",
       "GSM4137735  18.440538    21.608184       0.809690     0.976271  85.094543   \n",
       "GSM4137736  21.025726    26.555988       1.246243     1.177747  87.425606   \n",
       "GSM4137737  20.404150    23.738753       1.298100     1.171406  87.339378   \n",
       "GSM4137738  19.963135    21.407858       0.956814     1.199306  86.457207   \n",
       "GSM4137739  30.701977    25.379519       1.268822     1.193359  88.006493   \n",
       "GSM4137740  19.709518    26.994544       1.140624     1.174096  87.551102   \n",
       "GSM4137741  25.006363    29.815408       2.834759     1.106260  86.890930   \n",
       "GSM4137742  19.827387    29.370889       0.959647     1.257892  86.666138   \n",
       "GSM4137743  16.928745    27.138979       1.100993     1.179078  87.818886   \n",
       "GSM4137744  16.565306    23.659621       1.083474     1.186454  86.381805   \n",
       "GSM4137745  20.054005    31.345807       1.403846     1.169119  86.564972   \n",
       "\n",
       "            skinandblood  mammalian2  \n",
       "GSM4137709      2.354882    0.007100  \n",
       "GSM4137710      1.060467    0.007577  \n",
       "GSM4137711      1.289146    0.003947  \n",
       "GSM4137712      0.595460    0.003872  \n",
       "GSM4137713      0.786787    0.004309  \n",
       "GSM4137714      0.509315    0.003994  \n",
       "GSM4137715      0.786464    0.004123  \n",
       "GSM4137716      0.336474    0.004455  \n",
       "GSM4137717      0.562296    0.004236  \n",
       "GSM4137718      0.730006    0.004447  \n",
       "GSM4137719      2.630242    0.005451  \n",
       "GSM4137720      0.525785    0.004212  \n",
       "GSM4137721      0.459680    0.002872  \n",
       "GSM4137722      0.604637    0.004160  \n",
       "GSM4137723      0.853337    0.003363  \n",
       "GSM4137724      0.352591    0.004083  \n",
       "GSM4137725      0.660432    0.003064  \n",
       "GSM4137726      0.449509    0.003430  \n",
       "GSM4137727      0.565050    0.004039  \n",
       "GSM4137728      0.385037    0.005384  \n",
       "GSM4137729      0.580252    0.005072  \n",
       "GSM4137730      1.592233    0.008719  \n",
       "GSM4137731      0.346334    0.004748  \n",
       "GSM4137732      0.533043    0.003144  \n",
       "GSM4137733      0.343414    0.003209  \n",
       "GSM4137734      0.417019    0.004968  \n",
       "GSM4137735      0.386166    0.004448  \n",
       "GSM4137736      0.578503    0.003434  \n",
       "GSM4137737      0.717524    0.006604  \n",
       "GSM4137738      0.574664    0.003424  \n",
       "GSM4137739      0.663944    0.003469  \n",
       "GSM4137740      0.629927    0.004361  \n",
       "GSM4137741      1.303225    0.004899  \n",
       "GSM4137742      0.374429    0.004270  \n",
       "GSM4137743      0.499079    0.004071  \n",
       "GSM4137744      0.481675    0.003689  \n",
       "GSM4137745      0.473375    0.005378  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7d2d10dc-4ffe-4940-a7f1-2041b933f7b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'species': 'Homo sapiens',\n",
       " 'data_type': 'methylation',\n",
       " 'year': 2022,\n",
       " 'citation': 'de Lima Camillo, Lucas Paulo, Louis R. Lapierre, and Ritambhara Singh. \"A pan-tissue DNA-methylation epigenetic clock based on deep learning.\" npj Aging 8.1 (2022): 4.',\n",
       " 'doi': 'https://doi.org/10.1038/s41514-022-00085-y'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns['altumage_metadata']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "379e3a81-cafb-46b0-9d1a-313bd174ffa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pandas as pd\n",
    "#import matplotlib.pyplot as plt\n",
    "#from sklearn.decomposition import PCA\n",
    "#from sklearn.preprocessing import StandardScaler\n",
    "#\n",
    "## Assuming df is your DataFrame\n",
    "## Standardize the data\n",
    "#scaler = StandardScaler()\n",
    "#df_scaled = scaler.fit_transform(adata.obs)\n",
    "#\n",
    "## Create a PCA instance and fit_transform the scaled data\n",
    "#pca = PCA(n_components=2)\n",
    "#principal_components = pca.fit_transform(df_scaled)\n",
    "#\n",
    "## Convert to a DataFrame\n",
    "#pca_df = pd.DataFrame(data=principal_components, columns=['PC1', 'PC2'])\n",
    "#\n",
    "## Plotting\n",
    "#plt.figure(figsize=(8, 6))\n",
    "#plt.scatter(pca_df['PC1'], pca_df['PC2'])\n",
    "#plt.title('PCA of the Dataset')\n",
    "#plt.xlabel('Principal Component 1')\n",
    "#plt.ylabel('Principal Component 2')\n",
    "#plt.grid(True)\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a016ef7d-cf07-4e73-91f8-fe0884b706bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pca.explained_variance_ratio_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "eb9875b5-70b5-4fd9-a200-5de0c525c6f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the loadings for PC1\n",
    "#loadings = pca.components_[0]\n",
    "\n",
    "# Create a Series with the loadings and the original variable names\n",
    "#loading_series = pd.Series(loadings, index=adata.obs.columns)\n",
    "\n",
    "# Sort the loadings to see which variables contribute most to PC1\n",
    "#sorted_loading_series = loading_series.abs().sort_values(ascending=False)\n",
    "\n",
    "#print(\"Variables sorted by their contribution to PC1:\")\n",
    "#print(sorted_loading_series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "de7263da-ca0c-4c6c-af99-f6c04ae00ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pandas as pd\n",
    "#import numpy as np\n",
    "#import matplotlib.pyplot as plt\n",
    "#import seaborn as sns\n",
    "#from sklearn.decomposition import PCA\n",
    "#from sklearn.preprocessing import StandardScaler\n",
    "#\n",
    "## Example DataFrame\n",
    "## df = pd.read_csv('your_data.csv') # Uncomment this if you have a CSV file\n",
    "#\n",
    "## Example: Creating a sample DataFrame\n",
    "#np.random.seed(0)\n",
    "#df = adata.obs.copy()\n",
    "#\n",
    "## Standardizing the Data\n",
    "#scaler = StandardScaler()\n",
    "#df_scaled = scaler.fit_transform(df)\n",
    "#\n",
    "## Applying PCA\n",
    "#pca = PCA(n_components=min(df.shape))\n",
    "#pca.fit(df_scaled)\n",
    "#\n",
    "## Extracting Loadings\n",
    "#loadings = pd.DataFrame(pca.components_.T, columns=[f'PC{i+1}' for i in range(len(pca.components_))], index=df.columns)\n",
    "#\n",
    "## Plotting Heatmap of Loadings\n",
    "#plt.figure(figsize=(14, 10))\n",
    "#sns.heatmap(loadings, annot=True, cmap='coolwarm')\n",
    "#plt.title('PCA Loadings')\n",
    "#plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "948a0222-e6b5-4e0a-8176-4c2780332301",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import seaborn as sns\n",
    "#import matplotlib.pyplot as plt\n",
    "#\n",
    "## Assuming df is your DataFrame\n",
    "## Replace 'df' with your actual DataFrame name\n",
    "#corr_matrix = adata.obs.corr()\n",
    "#\n",
    "#corr_matrix=corr_matrix\n",
    "#\n",
    "## Create a clustermap\n",
    "#sns.clustermap(corr_matrix, annot=True, fmt=\".2f\", cmap='coolwarm', \n",
    "#               linewidths=0.5, figsize=(12, 12), dendrogram_ratio=(.15, .15),\n",
    "#               cbar_pos=(-0.05, .35, .03, .4))\n",
    "#\n",
    "## Add title\n",
    "#plt.title('')\n",
    "#\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d444a24e-6a98-4db1-8688-7f3f80ed2876",
   "metadata": {},
   "source": [
    "# Histone marks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dd281360-7e16-45d9-ae2b-8f8f3fff809d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyaging as pya"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b3988dd-3e39-4e13-8ad8-a06447137c52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting download_example_data function\n",
      "|-----------> ⚙️ Download example data started\n",
      "|-----------------> Data found in ./pyaging_data/ENCFF386QWG.bigWig\n",
      "|-----------> ✅ Download example data finished [0.0026s]\n",
      "|-----> 🎉 Done! [0.0053s]\n"
     ]
    }
   ],
   "source": [
    "pya.data.download_example_data('histone_mark')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e56f4080-e968-44cb-8e31-4bd27092f4b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting bigwig_to_df function\n",
      "|-----> ⚙️ Load Ensembl genome metadata started\n",
      "|-----------> Data found in ./pyaging_data/Ensembl-105-EnsDb-for-Homo-sapiens-genes.csv\n",
      "|-----> ✅ Load Ensembl genome metadata finished [0.0012s]\n",
      "|-----> ⚙️ Processing bigWig files started\n",
      "|-----------> Processing file: pyaging_data/ENCFF386QWG.bigWig\n",
      "|-----------> in progress: 100.0000%\n",
      "|-----------> Processing file: pyaging_data/ENCFF386QWG.bigWig\n",
      "|-----------> in progress: 100.0000%\n",
      "|-----> ✅ Processing bigWig files finished [0.2338s]\n",
      "|-----> 🎉 Done! [23.5367s]\n"
     ]
    }
   ],
   "source": [
    "df = pya.pp.bigwig_to_df(['pyaging_data/ENCFF386QWG.bigWig', 'pyaging_data/ENCFF386QWG.bigWig'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c501ed4c-f711-44be-bda4-669fbbae88f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting df_to_adata function\n",
      "|-----> ⚙️ Impute missing values started\n",
      "|-----------> No missing values found. No imputation necessary\n",
      "|-----> ✅ Impute missing values finished [0.0013s]\n",
      "|-----> ⚙️ Log data statistics started\n",
      "|-----------> There are 2 observations\n",
      "|-----------> There are 62241 features\n",
      "|-----------> Total missing values: 0\n",
      "|-----------> Percentage of missing values: 0.00%\n",
      "|-----> ✅ Log data statistics finished [0.0018s]\n",
      "|-----> ⚙️ Create anndata object started\n",
      "|-----> ✅ Create anndata object finished [0.0037s]\n",
      "|-----> ⚙️ Add metadata to anndata started\n",
      "|-----------? No metadata provided. Leaving adata.obs empty\n",
      "|-----> ⚠️ Add metadata to anndata finished [0.0072s]\n",
      "|-----> ⚙️ Add unstructured data to anndata started\n",
      "|-----> ✅ Add unstructured data to anndata finished [0.0010s]\n",
      "|-----> 🎉 Done! [0.0152s]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lucascamillo/mambaforge/envs/brain/lib/python3.9/site-packages/anndata/_core/anndata.py:1838: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = pya.preprocess.df_to_adata(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "13e7d357-c6d5-474b-b13f-30ec5edc0d19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting predict_age function\n",
      "|-----> ⚙️ Set PyTorch device started\n",
      "|-----------> Using device: cpu\n",
      "|-----> ✅ Set PyTorch device finished [0.0005s]\n",
      "|-----> Processing clock: h3k4me3\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Downloading data to ./pyaging_data/h3k4me3.pt\n",
      "|-----------------> in progress: 100.0000%\n",
      "|-----------> ✅ Load clock finished [0.0012s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [1.4113s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0021s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0036s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0044s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0040s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0010s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0014s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0022s]\n",
      "|-----> Processing clock: h3k9me3\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Downloading data to ./pyaging_data/h3k9me3.pt\n",
      "|-----------------> in progress: 100.0000%\n",
      "|-----------> ✅ Load clock finished [0.0007s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.9817s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0011s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0013s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0012s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0008s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0006s]\n",
      "|-----------> ⚙️ Load all clock metadata started\n",
      "|-----------------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----------> ✅ Load all clock metadata finished [0.0009s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0014s]\n",
      "|-----> 🎉 Done! [2.4223s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.pred.predict_age(adata, ['h3k4me3', 'h3k9me3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f64fb182-937b-4f67-b58e-5fffb0e2fad0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>h3k4me3</th>\n",
       "      <th>h3k9me3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pyaging_data/ENCFF386QWG.bigWig</th>\n",
       "      <td>53.998566</td>\n",
       "      <td>44.322887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pyaging_data/ENCFF386QWG.bigWig</th>\n",
       "      <td>53.998566</td>\n",
       "      <td>44.322887</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   h3k4me3    h3k9me3\n",
       "pyaging_data/ENCFF386QWG.bigWig  53.998566  44.322887\n",
       "pyaging_data/ENCFF386QWG.bigWig  53.998566  44.322887"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2946393e-a199-46ba-a9dd-80bc8fa88787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'species': 'Homo sapiens',\n",
       " 'data_type': 'histone_mark',\n",
       " 'year': 2023,\n",
       " 'citation': 'de Lima Camillo, Lucas Paulo, et al. \"Histone mark age of human tissues and cells.\" bioRxiv (2023): 2023-08.',\n",
       " 'doi': 'https://doi.org/10.1101/2023.08.21.554165'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns['h3k4me3_metadata']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf837dcf-de19-46f7-9c81-c063a45b14b6",
   "metadata": {},
   "source": [
    "# ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e64f6bb6-7c95-4b9e-b37a-4ae811dc088d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyaging as pya"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f94c6816-cb92-42e9-893b-b206def28922",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting download_example_data function\n",
      "|-----------> ⚙️ Download example data started\n",
      "|-----------------> Data found in ./pyaging_data/atac_example.pkl\n",
      "|-----------> ✅ Download example data finished [0.0041s]\n",
      "|-----> 🎉 Done! [0.0050s]\n"
     ]
    }
   ],
   "source": [
    "pya.data.download_example_data('atac')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "078b2b96-7317-4f84-a8c4-16276fb76137",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle('pyaging_data/atac_example.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "96be8dbe-9f71-461c-8cee-9cb7263b02b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak1</th>\n",
       "      <th>peak2</th>\n",
       "      <th>peak3</th>\n",
       "      <th>peak4</th>\n",
       "      <th>peak5</th>\n",
       "      <th>peak6</th>\n",
       "      <th>peak7</th>\n",
       "      <th>peak8</th>\n",
       "      <th>peak9</th>\n",
       "      <th>peak10</th>\n",
       "      <th>...</th>\n",
       "      <th>peak80391</th>\n",
       "      <th>peak80392</th>\n",
       "      <th>peak80393</th>\n",
       "      <th>peak80394</th>\n",
       "      <th>peak80395</th>\n",
       "      <th>peak80396</th>\n",
       "      <th>peak80397</th>\n",
       "      <th>peak80398</th>\n",
       "      <th>peak80399</th>\n",
       "      <th>peak80400</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sample_1</th>\n",
       "      <td>2.506150</td>\n",
       "      <td>4.042524</td>\n",
       "      <td>1.728068</td>\n",
       "      <td>1.779619</td>\n",
       "      <td>1.903873</td>\n",
       "      <td>3.513372</td>\n",
       "      <td>1.703487</td>\n",
       "      <td>3.553081</td>\n",
       "      <td>1.385296</td>\n",
       "      <td>2.069604</td>\n",
       "      <td>...</td>\n",
       "      <td>1.356951</td>\n",
       "      <td>1.197689</td>\n",
       "      <td>0.860598</td>\n",
       "      <td>0.916377</td>\n",
       "      <td>1.791575</td>\n",
       "      <td>0.834970</td>\n",
       "      <td>0.927765</td>\n",
       "      <td>1.504159</td>\n",
       "      <td>0.256097</td>\n",
       "      <td>2.776791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_2</th>\n",
       "      <td>3.281653</td>\n",
       "      <td>5.026707</td>\n",
       "      <td>2.903434</td>\n",
       "      <td>2.715392</td>\n",
       "      <td>2.723345</td>\n",
       "      <td>4.118045</td>\n",
       "      <td>2.952263</td>\n",
       "      <td>4.604813</td>\n",
       "      <td>3.853380</td>\n",
       "      <td>2.884414</td>\n",
       "      <td>...</td>\n",
       "      <td>2.934180</td>\n",
       "      <td>2.950956</td>\n",
       "      <td>2.696654</td>\n",
       "      <td>3.352670</td>\n",
       "      <td>1.086730</td>\n",
       "      <td>2.097016</td>\n",
       "      <td>2.440385</td>\n",
       "      <td>1.211147</td>\n",
       "      <td>2.129717</td>\n",
       "      <td>1.512977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_3</th>\n",
       "      <td>4.720936</td>\n",
       "      <td>6.620186</td>\n",
       "      <td>3.107523</td>\n",
       "      <td>4.066536</td>\n",
       "      <td>4.247777</td>\n",
       "      <td>6.454266</td>\n",
       "      <td>2.705629</td>\n",
       "      <td>5.671079</td>\n",
       "      <td>5.065204</td>\n",
       "      <td>3.530241</td>\n",
       "      <td>...</td>\n",
       "      <td>3.928536</td>\n",
       "      <td>3.422247</td>\n",
       "      <td>3.452730</td>\n",
       "      <td>1.294518</td>\n",
       "      <td>2.794667</td>\n",
       "      <td>2.957758</td>\n",
       "      <td>2.014624</td>\n",
       "      <td>2.608965</td>\n",
       "      <td>3.319312</td>\n",
       "      <td>3.156280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_4</th>\n",
       "      <td>4.690143</td>\n",
       "      <td>7.267509</td>\n",
       "      <td>4.832635</td>\n",
       "      <td>4.877077</td>\n",
       "      <td>5.082491</td>\n",
       "      <td>6.343680</td>\n",
       "      <td>4.185005</td>\n",
       "      <td>6.685465</td>\n",
       "      <td>6.565930</td>\n",
       "      <td>4.083189</td>\n",
       "      <td>...</td>\n",
       "      <td>2.552844</td>\n",
       "      <td>4.735987</td>\n",
       "      <td>6.339911</td>\n",
       "      <td>4.954476</td>\n",
       "      <td>2.692502</td>\n",
       "      <td>4.046954</td>\n",
       "      <td>5.168552</td>\n",
       "      <td>4.076252</td>\n",
       "      <td>3.276617</td>\n",
       "      <td>4.289285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_5</th>\n",
       "      <td>5.939011</td>\n",
       "      <td>8.126846</td>\n",
       "      <td>5.553807</td>\n",
       "      <td>5.824309</td>\n",
       "      <td>4.677195</td>\n",
       "      <td>8.828315</td>\n",
       "      <td>5.408961</td>\n",
       "      <td>7.399883</td>\n",
       "      <td>6.964666</td>\n",
       "      <td>5.146428</td>\n",
       "      <td>...</td>\n",
       "      <td>6.011554</td>\n",
       "      <td>4.259724</td>\n",
       "      <td>5.004213</td>\n",
       "      <td>7.080201</td>\n",
       "      <td>6.204265</td>\n",
       "      <td>3.588729</td>\n",
       "      <td>6.589236</td>\n",
       "      <td>5.816404</td>\n",
       "      <td>5.514156</td>\n",
       "      <td>6.268410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 80400 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             peak1     peak2     peak3     peak4     peak5     peak6  \\\n",
       "Sample_1  2.506150  4.042524  1.728068  1.779619  1.903873  3.513372   \n",
       "Sample_2  3.281653  5.026707  2.903434  2.715392  2.723345  4.118045   \n",
       "Sample_3  4.720936  6.620186  3.107523  4.066536  4.247777  6.454266   \n",
       "Sample_4  4.690143  7.267509  4.832635  4.877077  5.082491  6.343680   \n",
       "Sample_5  5.939011  8.126846  5.553807  5.824309  4.677195  8.828315   \n",
       "\n",
       "             peak7     peak8     peak9    peak10  ...  peak80391  peak80392  \\\n",
       "Sample_1  1.703487  3.553081  1.385296  2.069604  ...   1.356951   1.197689   \n",
       "Sample_2  2.952263  4.604813  3.853380  2.884414  ...   2.934180   2.950956   \n",
       "Sample_3  2.705629  5.671079  5.065204  3.530241  ...   3.928536   3.422247   \n",
       "Sample_4  4.185005  6.685465  6.565930  4.083189  ...   2.552844   4.735987   \n",
       "Sample_5  5.408961  7.399883  6.964666  5.146428  ...   6.011554   4.259724   \n",
       "\n",
       "          peak80393  peak80394  peak80395  peak80396  peak80397  peak80398  \\\n",
       "Sample_1   0.860598   0.916377   1.791575   0.834970   0.927765   1.504159   \n",
       "Sample_2   2.696654   3.352670   1.086730   2.097016   2.440385   1.211147   \n",
       "Sample_3   3.452730   1.294518   2.794667   2.957758   2.014624   2.608965   \n",
       "Sample_4   6.339911   4.954476   2.692502   4.046954   5.168552   4.076252   \n",
       "Sample_5   5.004213   7.080201   6.204265   3.588729   6.589236   5.816404   \n",
       "\n",
       "          peak80399  peak80400  \n",
       "Sample_1   0.256097   2.776791  \n",
       "Sample_2   2.129717   1.512977  \n",
       "Sample_3   3.319312   3.156280  \n",
       "Sample_4   3.276617   4.289285  \n",
       "Sample_5   5.514156   6.268410  \n",
       "\n",
       "[5 rows x 80400 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b90cf1f9-1179-4e56-a797-d0faecdea2ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting df_to_adata function\n",
      "|-----> ⚙️ Impute missing values started\n",
      "|-----------> No missing values found. No imputation necessary\n",
      "|-----> ✅ Impute missing values finished [0.0021s]\n",
      "|-----> ⚙️ Log data statistics started\n",
      "|-----------> There are 10 observations\n",
      "|-----------> There are 80400 features\n",
      "|-----------> Total missing values: 0\n",
      "|-----------> Percentage of missing values: 0.00%\n",
      "|-----> ✅ Log data statistics finished [0.0054s]\n",
      "|-----> ⚙️ Create anndata object started\n",
      "|-----> ✅ Create anndata object finished [0.0058s]\n",
      "|-----> ⚙️ Add metadata to anndata started\n",
      "|-----------? No metadata provided. Leaving adata.obs empty\n",
      "|-----> ⚠️ Add metadata to anndata finished [0.0035s]\n",
      "|-----> ⚙️ Add unstructured data to anndata started\n",
      "|-----> ✅ Add unstructured data to anndata finished [0.0021s]\n",
      "|-----> 🎉 Done! [0.0198s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.preprocess.df_to_adata(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "26398785-d1ea-4ce8-b1d9-7234f8f46ef6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting predict_age function\n",
      "|-----> ⚙️ Set PyTorch device started\n",
      "|-----------> Using device: cpu\n",
      "|-----> ✅ Set PyTorch device finished [0.0045s]\n",
      "|-----> Processing clock: ocampoatac1\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Downloading data to ./pyaging_data/ocampoatac1.pt\n",
      "|-----------------> in progress: 100.0000%\n",
      "|-----------> ✅ Load clock finished [0.0031s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [0.6640s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0024s]\n",
      "|-----------> ⚙️ Preprocess data started\n",
      "|-----------------> Preprocessing data with function log1p\n",
      "|-----------> ✅ Preprocess data finished [0.0045s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0049s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0037s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0020s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0054s]\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0028s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0035s]\n",
      "|-----> Processing clock: ocampoatac2\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Downloading data to ./pyaging_data/ocampoatac2.pt\n",
      "|-----------------> in progress: 100.0000%\n",
      "|-----------> ✅ Load clock finished [0.0019s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [4.3916s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0035s]\n",
      "|-----------> ⚙️ Preprocess data started\n",
      "|-----------------> Preprocessing data with function log1p\n",
      "|-----------> ✅ Preprocess data finished [0.0037s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0018s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0024s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0030s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0032s]\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0034s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0049s]\n",
      "|-----> 🎉 Done! [5.1219s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.pred.predict_age(adata, ['ocampoatac1', 'ocampoatac2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "cf13c211-d5ff-4aea-b1e5-7590869e3fb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ocampoatac1</th>\n",
       "      <th>ocampoatac2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sample_1</th>\n",
       "      <td>40.691948</td>\n",
       "      <td>39.464195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_2</th>\n",
       "      <td>40.281929</td>\n",
       "      <td>36.662262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_3</th>\n",
       "      <td>38.996227</td>\n",
       "      <td>34.257324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_4</th>\n",
       "      <td>39.658382</td>\n",
       "      <td>34.252609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_5</th>\n",
       "      <td>38.929848</td>\n",
       "      <td>33.717129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_6</th>\n",
       "      <td>39.158836</td>\n",
       "      <td>32.671177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_7</th>\n",
       "      <td>37.570061</td>\n",
       "      <td>31.182526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_8</th>\n",
       "      <td>37.766663</td>\n",
       "      <td>30.412004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_9</th>\n",
       "      <td>37.683392</td>\n",
       "      <td>29.650864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_10</th>\n",
       "      <td>37.072762</td>\n",
       "      <td>29.364052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ocampoatac1  ocampoatac2\n",
       "Sample_1     40.691948    39.464195\n",
       "Sample_2     40.281929    36.662262\n",
       "Sample_3     38.996227    34.257324\n",
       "Sample_4     39.658382    34.252609\n",
       "Sample_5     38.929848    33.717129\n",
       "Sample_6     39.158836    32.671177\n",
       "Sample_7     37.570061    31.182526\n",
       "Sample_8     37.766663    30.412004\n",
       "Sample_9     37.683392    29.650864\n",
       "Sample_10    37.072762    29.364052"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "6b368506-55d1-4b74-be61-817bcf575ade",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'species': 'Homo sapiens',\n",
       " 'data_type': 'atac',\n",
       " 'year': 2023,\n",
       " 'citation': 'Morandini, Francesco, et al. \"ATAC-clock: An aging clock based on chromatin accessibility.\" GeroScience (2023): 1-18.',\n",
       " 'doi': 'https://doi.org/10.1007/s11357-023-00986-0'}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns['ocampoatac1_metadata']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31cf37ce-09ee-49d7-a411-719bf65e186e",
   "metadata": {},
   "source": [
    "# Transcriptomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ad192191-e44f-4994-80ad-ab16cdb7c7e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyaging as pya"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "261fdc45-2b9b-407c-a49c-f045a750831d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting download_example_data function\n",
      "|-----------> ⚙️ Download example data started\n",
      "|-----------------> Data found in ./pyaging_data/GSE65765_CPM.pkl\n",
      "|-----------> ✅ Download example data finished [0.0075s]\n",
      "|-----> 🎉 Done! [0.0078s]\n"
     ]
    }
   ],
   "source": [
    "pya.data.download_example_data('rnaseq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "13aeb69a-4b0e-40f2-8094-194c9a6b42a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle('pyaging_data/GSE65765_CPM.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7af12fc3-1418-49df-ba7f-e94730db706e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WBGene00197333</th>\n",
       "      <th>WBGene00198386</th>\n",
       "      <th>WBGene00015153</th>\n",
       "      <th>WBGene00002061</th>\n",
       "      <th>WBGene00255704</th>\n",
       "      <th>WBGene00235314</th>\n",
       "      <th>WBGene00001177</th>\n",
       "      <th>WBGene00169236</th>\n",
       "      <th>WBGene00219784</th>\n",
       "      <th>WBGene00015152</th>\n",
       "      <th>...</th>\n",
       "      <th>WBGene00010964</th>\n",
       "      <th>WBGene00014467</th>\n",
       "      <th>WBGene00014468</th>\n",
       "      <th>WBGene00014469</th>\n",
       "      <th>WBGene00014470</th>\n",
       "      <th>WBGene00010965</th>\n",
       "      <th>WBGene00014471</th>\n",
       "      <th>WBGene00010966</th>\n",
       "      <th>WBGene00010967</th>\n",
       "      <th>WBGene00014473</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1793993</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.780174</td>\n",
       "      <td>169.240815</td>\n",
       "      <td>1.907427</td>\n",
       "      <td>0.277444</td>\n",
       "      <td>59.320986</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.283178</td>\n",
       "      <td>...</td>\n",
       "      <td>858.949156</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.052021</td>\n",
       "      <td>234.526846</td>\n",
       "      <td>0.017340</td>\n",
       "      <td>54.483057</td>\n",
       "      <td>78.117815</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1793991</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.510354</td>\n",
       "      <td>412.628597</td>\n",
       "      <td>0.061861</td>\n",
       "      <td>0.061861</td>\n",
       "      <td>22.239044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015465</td>\n",
       "      <td>0.201048</td>\n",
       "      <td>...</td>\n",
       "      <td>1049.982885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015465</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015465</td>\n",
       "      <td>372.511713</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>54.545971</td>\n",
       "      <td>59.618577</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1793994</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.718708</td>\n",
       "      <td>274.733671</td>\n",
       "      <td>1.234644</td>\n",
       "      <td>0.118391</td>\n",
       "      <td>42.400721</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.642691</td>\n",
       "      <td>...</td>\n",
       "      <td>664.255412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.101478</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>253.220421</td>\n",
       "      <td>0.033826</td>\n",
       "      <td>19.483698</td>\n",
       "      <td>86.492735</td>\n",
       "      <td>0.016913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1793992</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.389905</td>\n",
       "      <td>351.612558</td>\n",
       "      <td>0.505892</td>\n",
       "      <td>0.069778</td>\n",
       "      <td>20.497358</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017445</td>\n",
       "      <td>1.308342</td>\n",
       "      <td>...</td>\n",
       "      <td>1298.799849</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.034889</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>472.206803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>89.508039</td>\n",
       "      <td>76.459508</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 46755 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            WBGene00197333  WBGene00198386  WBGene00015153  WBGene00002061  \\\n",
       "SRR1793993             0.0             0.0        3.780174      169.240815   \n",
       "SRR1793991             0.0             0.0        0.510354      412.628597   \n",
       "SRR1793994             0.0             0.0        4.718708      274.733671   \n",
       "SRR1793992             0.0             0.0        2.389905      351.612558   \n",
       "\n",
       "            WBGene00255704  WBGene00235314  WBGene00001177  WBGene00169236  \\\n",
       "SRR1793993        1.907427        0.277444       59.320986             0.0   \n",
       "SRR1793991        0.061861        0.061861       22.239044             0.0   \n",
       "SRR1793994        1.234644        0.118391       42.400721             0.0   \n",
       "SRR1793992        0.505892        0.069778       20.497358             0.0   \n",
       "\n",
       "            WBGene00219784  WBGene00015152  ...  WBGene00010964  \\\n",
       "SRR1793993        0.000000        1.283178  ...      858.949156   \n",
       "SRR1793991        0.015465        0.201048  ...     1049.982885   \n",
       "SRR1793994        0.000000        0.642691  ...      664.255412   \n",
       "SRR1793992        0.017445        1.308342  ...     1298.799849   \n",
       "\n",
       "            WBGene00014467  WBGene00014468  WBGene00014469  WBGene00014470  \\\n",
       "SRR1793993             0.0        0.000000             0.0        0.052021   \n",
       "SRR1793991             0.0        0.015465             0.0        0.015465   \n",
       "SRR1793994             0.0        0.101478             0.0        0.000000   \n",
       "SRR1793992             0.0        0.034889             0.0        0.000000   \n",
       "\n",
       "            WBGene00010965  WBGene00014471  WBGene00010966  WBGene00010967  \\\n",
       "SRR1793993      234.526846        0.017340       54.483057       78.117815   \n",
       "SRR1793991      372.511713        0.000000       54.545971       59.618577   \n",
       "SRR1793994      253.220421        0.033826       19.483698       86.492735   \n",
       "SRR1793992      472.206803        0.000000       89.508039       76.459508   \n",
       "\n",
       "            WBGene00014473  \n",
       "SRR1793993        0.000000  \n",
       "SRR1793991        0.000000  \n",
       "SRR1793994        0.016913  \n",
       "SRR1793992        0.000000  \n",
       "\n",
       "[4 rows x 46755 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "23f7847f-8e33-4248-8273-6a48e9de31ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting df_to_adata function\n",
      "|-----> ⚙️ Impute missing values started\n",
      "|-----------> No missing values found. No imputation necessary\n",
      "|-----> ✅ Impute missing values finished [0.0032s]\n",
      "|-----> ⚙️ Log data statistics started\n",
      "|-----------> There are 4 observations\n",
      "|-----------> There are 46755 features\n",
      "|-----------> Total missing values: 0\n",
      "|-----------> Percentage of missing values: 0.00%\n",
      "|-----> ✅ Log data statistics finished [0.0041s]\n",
      "|-----> ⚙️ Create anndata object started\n",
      "|-----> ✅ Create anndata object finished [0.0043s]\n",
      "|-----> ⚙️ Add metadata to anndata started\n",
      "|-----------? No metadata provided. Leaving adata.obs empty\n",
      "|-----> ⚠️ Add metadata to anndata finished [0.0048s]\n",
      "|-----> ⚙️ Add unstructured data to anndata started\n",
      "|-----> ✅ Add unstructured data to anndata finished [0.0019s]\n",
      "|-----> 🎉 Done! [0.0182s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.preprocess.df_to_adata(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "18cde0cf-e977-4e0e-b42e-8f342728418a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-----> 🏗️ Starting predict_age function\n",
      "|-----> ⚙️ Set PyTorch device started\n",
      "|-----------> Using device: cpu\n",
      "|-----> ✅ Set PyTorch device finished [0.0021s]\n",
      "|-----> Processing clock: bitage\n",
      "|-----------> ⚙️ Load clock started\n",
      "|-----------------> Downloading data to ./pyaging_data/bitage.pt\n",
      "|-----------------> in progress: 100.0000%\n",
      "|-----------> ✅ Load clock finished [0.0044s]\n",
      "|-----------> ⚙️ Check features in adata started\n",
      "|-----------> All features are present in adata.var_names.\n",
      "|-----------> ✅ Check features in adata finished [3.2299s]\n",
      "|-----------> ⚙️ Convert adata.X to torch.tensor and filter features started\n",
      "|-----------> ✅ Convert adata.X to torch.tensor and filter features finished [0.0027s]\n",
      "|-----------> ⚙️ Preprocess data started\n",
      "|-----------------> Preprocessing data with function binarize\n",
      "|-----------> ✅ Preprocess data finished [0.0069s]\n",
      "|-----------> ⚙️ Initialize model started\n",
      "|-----------> ✅ Initialize model finished [0.0184s]\n",
      "|-----------> ⚙️ Predict ages with model started\n",
      "|-----------> ✅ Predict ages with model finished [0.0105s]\n",
      "|-----------> ⚙️ Convert tensor to numpy array started\n",
      "|-----------> ✅ Convert tensor to numpy array finished [0.0012s]\n",
      "|-----------> ⚙️ Add predicted ages to adata started\n",
      "|-----------> ✅ Add predicted ages to adata finished [0.0010s]\n",
      "|-----> ⚙️ Load all clock metadata started\n",
      "|-----------> Data found in ./pyaging_data/all_clock_metadata.pt\n",
      "|-----> ✅ Load all clock metadata finished [0.0015s]\n",
      "|-----------> ⚙️ Add clock metadata to adata.uns started\n",
      "|-----------> ✅ Add clock metadata to adata.uns finished [0.0027s]\n",
      "|-----> 🎉 Done! [3.2821s]\n"
     ]
    }
   ],
   "source": [
    "adata = pya.pred.predict_age(adata, ['bitage'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "032382f5-7d98-465e-a3cb-51165eeb7025",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bitage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1793993</th>\n",
       "      <td>182.353638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1793991</th>\n",
       "      <td>27.337227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1793994</th>\n",
       "      <td>241.629547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1793992</th>\n",
       "      <td>32.177971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                bitage\n",
       "SRR1793993  182.353638\n",
       "SRR1793991   27.337227\n",
       "SRR1793994  241.629547\n",
       "SRR1793992   32.177971"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "9908d25a-9639-4684-9da6-353c7eb4a555",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'species': 'C elegans',\n",
       " 'data_type': 'transcriptomics',\n",
       " 'year': 2021,\n",
       " 'citation': 'Meyer, David H., and Björn Schumacher. \"BiT age: A transcriptome‐based aging clock near the theoretical limit of accuracy.\" Aging cell 20.3 (2021): e13320.',\n",
       " 'doi': 'https://doi.org/10.1111/acel.13320'}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.uns['bitage_metadata']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
